File: continuedev/src/continuedev/plugins/steps/cmd.py
from textwrap import dedent
from typing import Coroutine

from ...core.main import Step
from ...core.observation import Observation
from ...core.sdk import ContinueSDK
from ...libs.util.strings import remove_quotes_and_escapes


class GenerateShellCommandStep(Step):
    user_input: str

    async def run(self, sdk: ContinueSDK) -> Coroutine[Observation, None, None]:
        cmd = await sdk.models.default.complete(
            dedent(
                f"""\
            The user has made a request to run a shell command. Their description of what it should do is:
            
            "{self.user_input}"

            Please write a shell command that will do what the user requested. Your output should consist of only the command itself, without any explanation or example output. Do not use any newlines. Only output the command that when inserted into the terminal will do precisely what was requested.
            """
            )
        )

        cmd = remove_quotes_and_escapes(cmd.strip()).replace("\n", "").replace("\r", "")

        await sdk.ide.runCommand(cmd)

        self.description = f"Generated shell command: {cmd}"

File: core/commands/slash/built-in-legacy/cmd.ts
import { streamLines } from "../../../diff/util.js";
import { SlashCommand } from "../../../index.js";
import { removeQuotesAndEscapes } from "../../../util/index.js";

function commandIsPotentiallyDangerous(command: string) {
  return (
    command.includes("rm -rf") ||
    command.includes("sudo") ||
    command.includes("cd / ")
  );
}

const GenerateTerminalCommand: SlashCommand = {
  name: "cmd",
  description: "Generate a shell command",
  run: async function* ({ ide, llm, input }) {
    if (input.trim() === "") {
      yield "Please provide a description of the shell command you want to generate. For example, '/cmd List all files in the current directory'.";
      return;
    }

    const gen = llm.streamComplete(
      `The user has made a request to run a shell command. Their description of what it should do is:

"${input}"

Please write a shell command that will do what the user requested. Your output should consist of only the command itself, without any explanation or example output. Do not use any newlines. Only output the command that when inserted into the terminal will do precisely what was requested. Here is the command:`,
      new AbortController().signal,
    );

    const lines = streamLines(gen);
    let cmd = "";
    for await (const line of lines) {
      console.log(line);
      if (line.startsWith("```") && line.endsWith("```")) {
        cmd = line.split(" ").slice(1).join(" ").slice(0, -3);
        break;
      }

      if (
        line.startsWith(">") ||
        line.startsWith("``") ||
        line.startsWith("\\begin{") ||
        line.trim() === ""
      ) {
        continue;
      }

      cmd = removeQuotesAndEscapes(line.trim());
      break;
    }

    yield `Generated shell command: ${cmd}`;
    if (commandIsPotentiallyDangerous(cmd)) {
      yield "\n\nWarning: This command may be potentially dangerous. Please double-check before pasting it in your terminal.";
    } else {
      await ide.runCommand(cmd);
    }
  },
};

export default GenerateTerminalCommand;
